accepted/2022/simplify-output-paths.md

```diff
@@ -0,0 +1,53 @@
+# Simplify .NET Output Paths
+
+We'd like to simplify the output paths for .NET 8.
+
+Currently, the default output path for a .NET project includes folders for the `TargetFramework` and the `RuntimeIdentifier` (if it is set).  Additionally, the publish output goes in a `publish` subfolder of the output folder.  The resulting paths are as follows:
+
+- `bin\<Configuration>\<TargetFramework>` - Build output with no `RuntimeIdentifier`
+- `bin\<Configuration>\<TargetFramework>\<RuntimeIdentifier>` - Build output with `RuntimeIdentifier`
+- `bin\<Configuration>\<TargetFramework>\publish` - Publish output with no `RuntimeIdentifier`
+- `bin\<Configuration>\<TargetFramework>\<RuntimeIdentifier>\publish` - Publish output with `RuntimeIdentifier`
+
+This makes it harder to find the output of a project, and sometimes nests one type of output in another output folder.
+
+## Proposed behavior
+
+Projects targeting .NET 8 and higher will by default not include the `TargetFramework` in the output path unless the project is multi-targeted.  Similarly, the `RuntimeIdentifier` will not by default be part of the output path.  Finally, the publish output path will be `bin\publish` by default instead of a `publish` subfolder inside of the output path.
```

Comment:
For mobile TFMs (-ios, -tvos) we need the RuntimeIdentifier in the path, because this is very common:

1. Build & test in the simulator (using a simulator RID)
2. At some pointer later, want to test on device (using a device RID).

Simulator and device builds are quite different, so not much can be reused between the RIDs (it would be a rather big task to figure out what can safely be re-used and make sure nothing else is). This means that changing between device and simulator when they build into the same location would have to invalidate the entire build (and then when you switch back the same thing would happen).

Another problem is that for multi-rid builds (say we're building for both `osx-arm64` and `osx-x64`), we build first for one RID, then the other RID, and at the end we merge the results into a universal app bundle. This wouldn't work if both RIDs built into the same directory.

So we'd probably have to set `AppendRuntimeIdentifierToOutputPath` to true always for any SDK for Apple platforms.

Another solution would be if the RID was only excluded from the output path of the project didn't specify a RuntimeIdentifier/RuntimeIdentifiers.

Reply:
I think that with this design, the iOS SDK would set the `AppendRuntimeIdentifierToOutputPath` to true.

In the future, I hope we can support multi-targeting over RuntimeIdentifiers.  For example you could set `TargetFrameworks` to something like `net9.0-macos/osx-x64;net9.0-macos/osx-arm64` and it would build both for you: https://github.com/dotnet/sdk/issues/9363#issuecomment-1238826925

Reply:
Maybe I am just being overly negative here, but if a whole set of platforms has to set `AppendRuntimeIdentifierToOutputPath=true` by default *anyway*, while others don't, it seems like this proposal in its current form adds very little value relative to the inconsistency and confusion it would cause for users.

Once again, I would point to the `pub\...` scheme I proposed above. It achieves the goals this proposal sets out to solve without any of that inconsistency and confusion.